11 feat/admin 관리자 CRUD 기능 및 Spring Security + JWT를 이용한 인증/인가 기능 구축

sshkim1013 opened the pull request:
## 관리자CRUD 기능 및 Spring Security + JWT를 이용한 인증/인가 기능 구축

변경된 파일이 굉장히 많아 각 클래스의 역할과 주요 로직을 정리한다고 정리했는데 읽기에 불편함이 없으실지 잘 모르겠습니다.
이렇게 한 번에 많은 파일을 올리게 된 점 다시 한 번 정말 죄송합니다... 😭 😭 😭

- **참고로 유림님께서 당시에 작성하셨던 Reservation 도메인의 Controller와 Service에서 Room 의존성을 못받아 에러가 발생했었는데 그걸 주석처리하고 push 할 때 주석을 해제한다는 걸 깜빡했습니다... 죄송합니다ㅠㅠ**

- **다들 꼭 Pull 받으시고 ReservationController와 ReservationService의 주석을 해제해주시면 감사하겠습니다 🙏 🙏**

<hr>

## 핵심 변경 사항 (Entity 수정)
### Admin 엔티티 PK 값 변경

- 기존 설계에서 `id`와 `userId`의 역할이 혼동될 수 있어 서로의 역할을 바꾸어 명확하게 변경했습니다.

  - 기존 -> `userId`(PK, 자동증가), `id`(로그인 ID)
  - **변경 -> `id`: DB의 Primary Key (자동 증가 Long 타입), `userId`: 실제 사용자가 로그인할 때 쓰는 계정 ID (String 타입, Unique)**
 
> 일반적으로 `id`라는 것은 DB 관리를 위한 고유 식별자로 명시되는 경우가 많기 때문에 기존처럼 하면 혼란을 일으킬 수 있다고 생각했습니다.

<hr>

## 주요 클래스 설명
### 1. Security & 인증 설정 (config 패키지)

### 1-1. `SecurityConfig.java`
**Spring Security의 문지기 설정**
  - CSRF 비활성화 및 Stateless 세션 정책 설정 (JWT 사용을 위해 필수)
  - `/admin/auth/**` 경로는 인증 없이 통과, 그 외는 인증 필요
  - 우리가 만든 JwtAuthenticationFilter를 필터 체인에 등록
  - `@EnableMethodSecurity` : 메서드 위에서 권한 체크(@PreAuthorize)를 할 수 있게 활성화

### 1-2. `JwtProvider.java`
 **JWT 토큰 공장 (생성, 검증, 파싱)**
  - Access Token (1시간) & Refresh Token (7일) 생성
  - 토큰 타입 구분: Access 토큰으로 Refresh 요청을 하는 것을 막기 위해 type 정보 포함
  - 블랙리스트 관리: 로그아웃된 토큰을 메모리에 저장하여 재사용 방지

### 1-3. `JwtAuthenticationFilter.java`
**들어오는 요청의 신분증(토큰) 검사원**
  - 요청 헤더(Authorization)에서 토큰 추출
  - 토큰이 유효하고 블랙리스트에 없다면 인증 정보(Authentication)를 생성해 `SecurityContext`에 저장
  - 유효하지 않은 토큰일 경우 401 Unauthorized 에러와 함께 명확한 JSON 메시지 응답

### 1-4. `InitialDataLoader.java`
**초기 MASTER 계정 생성기**
  - 서버 시작 시 MASTER 계정이 없으면 자동으로 생성
  - 최초 설정 : `userId`: master, `password`: password123 (개발 편의성)

<hr>

### 2. 인증 / 인가 파트(Auth)

### 2-1. `AuthController.java`
**회원가입과 로그인, 토큰 재발급과 같은 인증/인가 처리 로직의 진입점**

### 2-2. `AuthService.java`
**인증 관련 핵심 로직 처리**
- **로그인**: 비밀번호 확인 후 `Access / Refresh` 토큰 발급
- **로그아웃**: 토큰을 블랙리스트에 추가하여 즉시 무효화
- **토큰 갱신 (Refresh)**
  - **RTR(Refresh Token Rotation) 적용**: 갱신 시 기존 Refresh Token뿐만 아니라 기존 Access Token까지 모두 블랙리스트에 추가하여 보안 강화.

<hr>


### 3. 관리자 파트(Admin)

### 3-1. `AdminController.java`
**관리자 CRUD API**
- `@PreAuthorize`를 사용하여 권한별 접근 제어
- **MASTER**: 모든 관리자 조회 및 삭제 가능
- **ADMIN**: 본인 정보만 조회 및 수정 가능

### 3-2. `AdminService.java`
**관리자 관리 로직**
- 관리자 조회 시 현재 로그인한 사용자의 권한을 체크하여 접근 제어
- 비밀번호 수정: **본인만 수정 가능**하도록 검증 로직 포함

<hr>

### 4. 예외 처리 (exception 패키지)
### 4-1. `GlobalExceptionHandler.java`
- 인증 실패(401), 권한 없음(403) 등 모든 에러를 가로채서 통일된 JSON 형식({ "message": "..." })으로 반환

### 4-2. `ErrorCode.java`
에러 추가: `TOKEN_INVALID`, `ADMIN_ACCESS_DENIED` 등 상황별 에러 정의

<hr>

### 5. 테스트 결과(Postman 테스트 사진 첨부)

5-1. **`SayyeApplication` 실행 시 최초에 MASTER 관리자 생성**

<img width="976" height="104" alt="첫 생성 유저 MASTER 등록" src="https://github.com/user-attachments/assets/c9b03c26-b170-436b-838b-4c9ec55c61ea" />


<br>**5-2. MASTER 관리자 로그인**

<img width="748" height="491" alt="MASTER 로그인" src="https://github.com/user-attachments/assets/77b0f7bc-1809-48bd-b39c-559ef9e6a0e6" />


<br>**5-3. 관리자 전체 조회 - MASTER 관리자만 가능**

<img width="1003" height="740" alt="관리자 전체 조회 - 마스터관리자만 가능" src="https://github.com/user-attachments/assets/edef9551-9598-4264-9f91-061fa8e69b95" />

<br>**5-4. 관리자 단일 조회 - MASTER 관리자와 해당 관리자 본인만 가능**

- MASTER 관리자가 2번 관리자 조회

<img width="1005" height="634" alt="관리자 단일 조회(마스터 관리자가 2번 유저 조회)" src="https://github.com/user-attachments/assets/6b00e9e7-4d86-4ef0-8371-75f7de6968a1" />

<br>**5-5. 리프레쉬 토큰 재발급 성공**

<img width="1004" height="607" alt="리프레쉬 토큰 재발급 성공" src="https://github.com/user-attachments/assets/97315925-68ba-44e1-9410-05bb27b39d08" />

<br>**5-6. 리프레쉬 토큰 재발급 후 기존의 accessToken을 사용하여 관리자 전체 조회 - 실패(정상 동작)**

<img width="1003" height="357" alt="리프레쉬 토큰 재발급 후 기존의 액세스토큰 사용(실패)" src="https://github.com/user-attachments/assets/06ec732a-5371-4f3d-be44-568cec1898ec" />

<br>**5-7. (본인의) 비밀번호 수정 성공**

<img width="1002" height="559" alt="본인의 비밀번호 수정" src="https://github.com/user-attachments/assets/2a19ae68-eb07-44db-b33b-4955171c74fd" />

<br>**5-8. 비밀번호 수정 후 기존의 비밀번호로 재로그인 시도 - 실패(정상 동작)**

<img width="1003" height="563" alt="비밀번호 수정 후 기존 비밀번호로 로그인(실패)" src="https://github.com/user-attachments/assets/d370a738-e0a3-4c02-95d6-dccc0214c8cf" />

<br>**5-9. 비밀번호 수정 후 변경된 비밀번호로 로그인 성공**

<img width="999" height="609" alt="비밀번호 수정 후 재로그인 성공" src="https://github.com/user-attachments/assets/915c9ade-c715-40ec-ae61-a1adbe438295" />


<br>**5-10. 관리자 삭제 - MASTER 관리자만 가능**

<img width="1003" height="525" alt="관리자 삭제(MASTER만 가능)" src="https://github.com/user-attachments/assets/30bab956-b657-4839-bfa4-b7a0eafb2317" />


<br>**5-11. 1번 관리자 삭제 후 관리자 전체 조회 - 1번 관리자 삭제 완료**

<img width="1000" height="711" alt="1번 관리자 삭제 후 전체 조회" src="https://github.com/user-attachments/assets/ac6dc971-aacd-4786-a66d-a87e1f78c9ae" />


<br>**5-12. 로그아웃 성공**

<img width="1004" height="557" alt="로그아웃 성공" src="https://github.com/user-attachments/assets/b852382e-f7b9-450a-9c6e-88b9ac9255d9" />


<br>**5-13. 로그아웃 이후 다시 로그아웃 시도 - 실패(정상 동작)**

<img width="1003" height="557" alt="로그아웃 상태에서 다시 로그아웃(에러)" src="https://github.com/user-attachments/assets/1e13755c-106b-491b-b788-2063ce6f8e88" />

<br>**5-14. 새로운 관리자(ADMIN) 회원가입 성공**

<img width="1003" height="634" alt="새로운 관리자 회원가입(성공)" src="https://github.com/user-attachments/assets/aecf4e5d-49c3-4a8a-9645-8e9cc09aa75a" />


<br>**5-15. 기존 유저가 다시 회원가입 시도 - 실패(정상 동작)**

<img width="1001" height="557" alt="기존 유저가 다시 회원가입(실패)" src="https://github.com/user-attachments/assets/61d2151e-46f0-4b02-955a-98b55d77fe3a" />

<br>**5-16. 다른 유저의 비밀번호 수정 시도 - 실패(정상 동작)**

- 4번 유저가 3번 유저의 비밀번호 수정 시도

<img width="1005" height="554" alt="4번 유저가 3번 유저의 비밀번호 수정(에러)" src="https://github.com/user-attachments/assets/fe8c4e3a-454b-4fe9-b59f-bc444c17861c" />

<br>**5-17. MASTER 관리자가 아닌 관리자(ADMIN)가 다른 관리자 삭제 시도 - 실패(정상 동작)**

<img width="1000" height="556" alt="마스터가 아닌 관리자의 타 관리자 삭제(실패)" src="https://github.com/user-attachments/assets/b6e64d8e-8fda-47ea-9ec0-716151ae4d6b" />


<hr>

### 참고

Postman 테스트 시에 `Authorization` 탭에서 `AccessToken`을 입력해야 하는 테스트 목록은 다음과 같습니다.

1. 관리자 토큰 재발급 API (POST localhost:8080/admin/auth/refresh) 
2. 관리자 전체 조회(GET localhost:8080/admin/users) - MASTER 관리자만 가능
3. 관리자 단일 조회(GET localhost:8080/admin/users/{userId}) - MASTER 관리자와 해당 유저 본인만 가능
4. 관리자 비밀번호 수정(PATCH localhost:8080/admin/users/{userId}) - 해당 유저 본인만 가능
5. 관리자 삭제 (DELETE localhost:8080/admin/users/5) - MASTER 관리자만 가능


#### 일단 최대한 편하게 이해하실 수 있도록 비유적 표현도 추가해서 내용을 작성했는데 그럼에도 너무 많은 내용이다보니 불편하실 것이라 생각합니다. 정말 죄송합니다ㅜ.ㅜ 그래도 Postman을 통해 여러 번 작업을 했고, 특별히 문제가 발생하지 않음도 확인하였습니다! 천천히 확인해주시면 감사하겠습니다 😄 😄

sshkim1013 committed:
1125 SAYYE 관리자 파트 개발

sshkim1013 committed:
1126 변수명 수정
service -> adminService
repository -> adminRepository

sshkim1013 committed:
Merge branch 'dev' into 11-feat/admin

sshkim1013 committed:
Merge branch 'dev' into 11-feat/admin

sshkim1013 committed:
Merge branch 'dev' into 11-feat/admin

sshkim1013 committed:
feat/Admin(#25): Spring Security 및 JWT 인증 인프라 구축

sshkim1013 committed:
feat(#11): 관리자(Admin) 인증 및 CRUD 기능 구현

sshkim1013 committed:
feat/admin(#11): MASTER 관리자 초기 생성

ghkdquddlf reviewed (dismissed):
수고하셨습니다!

yu-rim98 commented on src/main/java/com/sayye/admin/config/JwtAuthenticationFilter.java:
```diff
@@ -0,0 +1,76 @@
+import java.util.Map;
+
+public class JwtAuthenticationFilter extends OncePerRequestFilter {
+    private final JwtProvider jwtProvider;
```
`@RequiredArgsConstructor` 어노테이션을 사용해서 생성자 부분을 지울 수 있을 것 같습니다.

yu-rim98 commented on src/main/java/com/sayye/admin/config/JwtAuthenticationFilter.java:
```diff
@@ -0,0 +1,76 @@
+        if (requestURI.startsWith("/admin/auth/")) {
+            chain.doFilter(request, response);
+            return;
+        }
```
`SecurityConfig`에 `.requestMatchers("/admin/auth/**").permitAll()` 동일한 경로가 설정되어 있어서 시큐리티에서 통과되지 않을까 싶은데 여기서도 설정해두신 이유가 궁금합니다.

sshkim1013 replied:
저도 처음엔 헷갈렸는데, `SecurityConfig`의 `.permitAll()` 메서드는 **인증 정보가 없어도 최종 통과시켜라**라는 뜻이지만, 필터 자체를 건너뛰라는 뜻은 아니더라구요.

`JwtAuthenticationFilter`는 기본적으로 모든 요청에 대해 토큰을 검사하고 없으면 권한 없음 에러(401)를 반환하도록 되어 있기 때문에 로그인 요청(/admin/auth/**)처럼 **토큰이 없는 요청이 들어오면** `permitAll()` 이 있는 곳까지 가보기도 전에 필터에서 401 에러가 발생해 버립니다.

그래서 필터 내부에서 `chain.doFilter()` 메서드를 통해 **회원가입이나 로그인 경로는 토큰 검사를 하지 말고 그냥 통과시켜라** 라고 명시적으로 길을 열어준 것입니다. 그래야 `SecurityConfig`의 `permitAll()`까지 요청이 전달될 수가 있습니다.

yu-rim98 replied:
오오 그렇군요 넵 감사합니다!

yu-rim98 commented on src/main/java/com/sayye/admin/config/JwtAuthenticationFilter.java:
```diff
@@ -0,0 +1,76 @@
+        
+        String header = request.getHeader("Authorization");
+
+        if (header != null && header.startsWith("Bearer ")) {
```
`Authorization`, `Bearer ` 처럼 하드코딩되어 있는 부분은 상수로 관리하면 더 좋을 것 같습니다.

yu-rim98 commented on src/main/java/com/sayye/admin/config/JwtProvider.java:
```diff
@@ -0,0 +1,99 @@
+public class JwtProvider {
+    // HS256 알고리즘은 최소 256비트(32바이트) 키가 필요
+    // production 환경에선 환경변수로 관리 권장
+    private final SecretKey secretKey = Keys.hmacShaKeyFor("sayyeSecretKeyForJWTTokenGenerationMustBeAtLeast32BytesLong".getBytes());
```
yml에 저장해서 불러오는 방식으로 사용하는 것이 더 좋을 것 같습니다.

yu-rim98 commented on src/main/java/com/sayye/admin/config/JwtProvider.java:
```diff
@@ -0,0 +1,99 @@
+                    .setSigningKey(secretKey)
+                    .build()
+                    .parseClaimsJws(token);
+            return true;
+        } catch (Exception e) {
+            return false;
+        }
```
토큰 검증 시 발생하는 예외마다 catch로 잡아서 로그를 남겨두면 어떤 에러인지 확인할 수 있을 것 같아서 좋을 것 같습니다. (잘못된 JWT, 만료된 JWT 등)

yu-rim98 commented on src/main/java/com/sayye/admin/controller/AdminController.java:
```diff
@@ -0,0 +1,65 @@
+
+@RestController
+@RequiredArgsConstructor
+@RequestMapping("/admin/users")
```
`admin`이 관리자라는 의미를 담고 있으니까 users는 붙이지 않고 `admins`로도 충분할 것 같습니다.

sshkim1013 replied:
초반 규약 맞추려다보니 이렇게 만들게 되었네요. `/admins` 로 축약하겠습니다!

yu-rim98 commented on src/main/java/com/sayye/admin/controller/AuthController.java:
```diff
@@ -0,0 +1,71 @@
+
+@RestController
+@RequiredArgsConstructor
+@RequestMapping("/admin/auth")
```
Auth는 인증 관련된 것만 담당하니까 `/auth`로 가셔도 좋을 것 같습니다.

yu-rim98 commented on src/main/java/com/sayye/admin/controller/AuthController.java:
```diff
@@ -0,0 +1,71 @@
+        
+        HttpHeaders headers = new HttpHeaders();
+        headers.set("Authorization", "Bearer " + tokens.getAccessToken());
+        headers.set("Refresh-Token", tokens.getRefreshToken());
```
이 부분은 아래 부분이랑 중복이 되는 것 같아요. 메서드로 묶을 수 있을 것 같습니다.

yu-rim98 commented on src/main/java/com/sayye/admin/controller/AuthController.java:
```diff
@@ -0,0 +1,71 @@
+    @PostMapping("/logout")
+    public ResponseEntity<LogoutResponse> logout(HttpServletRequest request) {
+        return ResponseEntity.ok(authService.logout(request));
+    }
```
`LogoutResponse`에는 "성공적으로 로그아웃 되었습니다." 문자열만 담는 것 같은데 DTO로 만드신 이유가 궁금합니다.
단순히 로그인 성공 했다는 문자열만 필요하다면 `ResponseEntity<String>`으로 반환하고 바디 안에 문자열을 직접 넣어줘도 괜찮을 것 같습니다. (`UpdatePasswordResponse`, `LoginResponse`도 동일)

sshkim1013 replied:
그렇네요.. DTO 세분화 생각하고 급하게 작업 하다보니 이렇게 설계해버린 것 같네요ㅠ

yu-rim98 commented on src/main/java/com/sayye/admin/dto/response/AdminResponse.java:
```diff
@@ -0,0 +1,30 @@
+import lombok.RequiredArgsConstructor;
+
+@Getter
+@RequiredArgsConstructor
```
`new AdminResponse(Admin admin)`를 통해서만 생성되고 있는 것 같아서 `RequiredArgsConstructor`는 제거해주셔도 될 것 같습니다.

yu-rim98 commented on src/main/java/com/sayye/admin/service/AdminService.java:
```diff
@@ -0,0 +1,108 @@
+
+        for (Admin admin : admins) {
+            dtoList.add(new AdminResponse(admin));
+        }
```
`stream`을 사용해서 코드를 조금 줄일 수 있을 것 같습니다.

yu-rim98 commented on src/main/java/com/sayye/admin/service/AdminService.java:
```diff
@@ -0,0 +1,108 @@
+                .orElseThrow(() -> new ApiException(ErrorCode.ADMIN_NOT_FOUND_ERROR));
+        
+        // MASTER는 모든 관리자 조회 가능, ADMIN은 본인만 조회 가능
+        if (currentAdmin.getRole() != Role.MASTER && !admin.getUserId().equals(currentUserId)) {
```
이 조건을 확인하는 로직을 엔티티의 메서드로 만들고 호출하는 방법도 좋을 것 같아요.

yu-rim98 commented on src/main/java/com/sayye/admin/service/AdminService.java:
```diff
@@ -0,0 +1,108 @@
+        admin.setPassword(passwordEncoder.encode(req.getPassword()));
+        admin.setName(req.getName());
+        admin.setEmail(req.getEmail());
+        admin.setRole(Role.ADMIN); // 기본 ADMIN 권한 부여, 추후 설정 가능
```
setter를 통해 엔티티의 값을 직접 수정하는 것보다 엔티티 생성 시 값을 넣어주는게 더 안전할 것 같습니다.
매개변수가 있는 생성자를 만들거나 정적 팩토리 메서드를 통해 객체를 생성하면 setter 제거 및 기존 엔티티 클래스에 `NoArgsConstructor` 접근 제어 `PROTECTED` 레벨을 유지할 수 있을 것 같아요.

yu-rim98 commented on src/main/java/com/sayye/admin/service/AuthService.java:
```diff
@@ -0,0 +1,158 @@
+    @Transactional
+    public TokenPair login(LoginRequest loginRequest) {
+        Admin admin = adminRepository.findByUserId(loginRequest.getUserId())
+                .orElseThrow(() -> new ApiException(ErrorCode.ADMIN_NOT_FOUND_ERROR));
```
AuthService는 어차피 AdminService를 주입받고 있는 것 같아 보입니다. AdminRepository를 직접 호출해서 예외 던지는 것까지 처리할 필요가 없이 AdminService의 메서드를 호출하는 방법도 괜찮을 것 같아요.

yu-rim98 reviewed (commented):
코멘트 확인 부탁드립니다! 고생 많으셨어요!

sshkim1013 committed:
refactor/admin(#11): JWT 설정 외부화 및 하드코딩 상수 제거

- application.yml: jwt.secret 설정 추가
- JwtProvider: AUTHORIZATION_HEADER, BEARER_PREFIX 상수 추가(하드코딩된 코드 대체)
  - "Authorization", "Bearer "
  - substring(7) -> substring(BEARER_PREFIX.length())

sshkim1013 committed:
refactor/admin(#11): Admin 엔티티에 정적 팩토리 메서드 적용

sshkim1013 committed:
feat/admin(#11): JWT 토큰 검증 예외 처리 및 JwtProvider 로깅 개선

sshkim1013 committed:
chore/admin(#11): 코드 정리 및 성능 개선

- InitialDataLoader 개선
- SecurityConfig 수정
- AdminController 리팩토링
- AdminResponse 정리

sshkim1013 committed:
refactor/admin(#11): 불필요한 DTO 클래스 제거

- LoginResponse
- LogoutResponse
- UpdatePasswordResponse

String 응답으로 대체

yu-rim98 commented on src/main/java/com/sayye/admin/service/AdminService.java:
```diff
@@ -0,0 +1,118 @@
+                .orElseThrow(() -> new ApiException(ErrorCode.ADMIN_NOT_FOUND_ERROR));
+        
+        // MASTER는 모든 관리자 조회 가능, ADMIN은 본인만 조회 가능
+        RoleAndSameUserVerification(currentUserId, currentAdmin, admin);
```
메서드명이 대문자로 시작하는 것 같습니다!

yu-rim98 commented on src/main/java/com/sayye/admin/service/AdminService.java:
```diff
@@ -0,0 +1,118 @@
+        }
+
+        // 정적 팩토리 메서드를 통한 객체 생성
+        Admin admin = Admin.createAdmin(
```
정적 팩토리 메서드는 다른 정적 메서드와 역할을 구분하기 위해 네이밍 규칙이 존재합니다.
여러 개의 매개 변수를 받아 객체를 생성하는 메서드의 경우 of를 사용하기 때문에 of로 수정해주시면 좋을 것 같습니다!

yu-rim98 commented on src/main/java/com/sayye/admin/service/AdminService.java:
```diff
@@ -0,0 +1,118 @@
+        if (currentAdmin.getRole() != Role.MASTER && !admin.getUserId().equals(currentUserId)) {
+            throw new ApiException(ErrorCode.ADMIN_ACCESS_DENIED);
+        }
+    }
```
개인적으로는 해당 메서드를 서비스에 두는 것보다 엔티티의 메서드로 두고 서비스에서 직접 엔티티의 데이터를 get으로 가져와서 검증하기 보다 엔티티에게 묻고 결과에 따라서 처리하는게 더 좋을 것 같아요. 개인적인 생각이라 꼭 반영하진 않으셔도 됩니다!

yu-rim98 reviewed (commented):
코멘트 확인 부탁드려요!

sshkim1013 committed:
Merge branch 'dev' into 11-feat/admin

sshkim1013 committed:
refactor/admin(#11): Admin 엔티티에 본인 검증 로직 메서드 추가 및 정적 팩토리 메서드명 수정

yu-rim98 reviewed (dismissed):
확인했습니다. 고생 많으셨습니다!!

ghkdquddlf reviewed (dismissed):
수고하셨습니다!

ohmyeongje6541 commented:
고생하셨습니다

sshkim1013 committed:
Merge branch 'dev' into 11-feat/admin
